### ISSUE TITLE
Perhaps Router Resolve shouldn't wait for Observable to Complete?

### ISSUE THREAD

**Issue by syndicatedshannon** [angular/angular#10556]
This is probably best classified as a Router feature request. I'll describe what it is, my thinking, and a use case.

What it is: Currently, the V3 Router waits for an Observable to Complete before completing navigation. I would like for it to instead continue navigation after the first Observable item is returned.

My thinking: Currently the Resolve treats the Observable as a Promise, defeating the power of the Observable. This doesn't present a barrier with Angular's Http, where only one item is returned. But unless I misunderstand, the point of using an Observable in Http is to allow easy extension throughout Angular into a scenario where the data is actually a stream.

Use case: I have a live data feed from a remote server. I subscribe to a resource, rather than get it. When viewing this resource in a page, I only need one snapshot to complete navigation and display the component, but I would still like the updates bound to the view.

More detail:

Observables as first-class citizens in Angular2 are great for displaying "live data". I currently rely on, and it is becoming more common to integrate, "live data" streamed/syndicated/published from web services. Regardless of the framework used to deliver streams to the client, Observables are the obvious choice to act as intermediary between the receiving agent/service and the view.

Resolvers in Angular are also very nice for a few reasons. One is that I can be sure I have all my data before performing initialization. This greatly simplifies my component initialization logic. It also simplifies managing built-in operations Angular performs, such as avoiding errors binding templates to properties on null models.

But where live data comes in, the current Resolve behavior is difficult to make use of. If we suppose the live data stream never completes, we have to instead resolve, for example, to Observable.first(), and throw away the remainder of the stream.

Or I can go through extra steps for each of my Observables, to Observable.publish etc. But that's not enough, because there's not a symmetric exit hook for resolve, that I see. So instead I have to do some extra magic, such as tear down resources I didn't initialize in the component.

IMO, in the spirit of Rx, a better behavior would be to:
- Share the Observable, e.g. via PublishReplay
- Connect
- Subscribe
- Wait for the first item
- Unsubscribe
- Complete navigation 
- Pass the shared Observable as ResolvedData
- When exiting navigation Disconnect

Or something similar.

Your thoughts are appreciated. I'll also be happy to contribute effort if someone with design authority confirms an approach.

**Comment by johnchristopherjones** [comment]
Actually been bashing my head against the wall for the last two days over this exact thing.   I had no idea the observable had to complete order for the navigation to complete.   Retrospectively, I understand it where it came from for  promise based approach.  Without live observables, it is proving to be difficult to merge the angular router with John Lindquist's course on observables: https://egghead.io/courses/building-a-time-machine-with-angular-2-and-rxjs

**Comment by syndicatedshannon** [comment]
@johnchristopherjones

Yes, although I wasn't thinking about this as a defect-inducing issue because we've moved past that, I know what you mean, and we did initially experience that confusion as well. Several times our team hit a multi-hour snag trying to track this same issue down. The first was me, the next two were the same team member hitting it twice, once after a refactor. Embarrassingly, we both still stared at it for a while the second time: As usual, locating a defect with no visible error can be time-consuming.

What is the workaround you are looking for? Yes, observable.take(1), a.k.a. observable.first() is how we initially hacked the issue. Unless I misunderstand your question, hot vs. cold not directly related.

However, observable.first() will throw away all the following data. That's only a problem if you care about the following data (which you probably do, since otherwise your observable would probably be "completed").

**Comment by syndicatedshannon** [comment]
@johnchristopherjones : Side note, I didn't see your question earlier. If you added it in an edit, you may want to post it separately next time.

**Comment by kemsky** [comment]
This is good idea, but i see at least two problems:
1. once router stops waiting for observable to complete we will loose possibility to do *redirect* before component is created. I would suggest to make Resolve configurable.
2. we need to unsubscribe at some point and it gets tricky when components are reused by the router.

**Comment by DzmitryShylovich** [comment]
I believe this was fixed https://github.com/angular/angular/pull/10412

**Comment by syndicatedshannon** [comment]
Thank you for the find/reference, Dzmitry.

I raised a specific issue in my "desired behavior":

- When exiting navigation Disconnect


Is there a pattern in-mind, on how disconnects are handled now, after #10412 ?

**Comment by syndicatedshannon** [comment]
@DzmitryShylovich After an errant discussion on #10412, it looks like it, although it is a related behavior, it has no effect on this issue.

That PR changes the behavior of guards (OnActivate), not Resolve.

**Comment by swftvsn** [comment]
Router 4 should really support hot routes, that is views that are updated once the observable emits new data. 

The activated route should flow the data and if route change is needed it should happen after the observable has emitted the *first* value.

The disconnect to the resolve observable should happen when the route is not activated anymore.

**Comment by greggbjensen** [comment]
I have this same problem.  I used a Resolve with a Promise for an Observable to work around it.  My service in this case caches the data.  It would be nice if this was supported natively.

```javascript
@Injectable()
export class DocumentsResolve implements Resolve<Promise<Observable<IDocument[]>>> {
    constructor(
        private _documentService: DocumentService,
    ) {}

    public resolve(
            route: ActivatedRouteSnapshot,
            state: RouterStateSnapshot): Promise<Observable<IDocument[]>> {

        return new Promise((resolve, reject) => {
            const documentObservable = this._documentService
                .list();

            documentObservable
                .first()
                .subscribe(() => resolve(documentObservable));
        });
    }
}
```

```javascript
@Component({
    selector: 'app-document-library',
    templateUrl: './document-library.component.html',
    styleUrls: ['./document-library.component.css']
})
export class DocumentLibraryComponent implements OnInit, OnDestroy {
    public documents: Observable<IDocument[]>;

    constructor(
        private _route: ActivatedRoute,
    ) {}

    public ngOnInit(): void {
        this.documents = this._route.snapshot.data['documents'];
    }
}
```

**Comment by greggbjensen** [comment]
Here is an example of simplifying above with a re-usable base class.

```javascript
export abstract class HotResolve<T extends Observable<any>> implements Resolve<Promise<T>> {
    public resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<T> {
        return new Promise((resolve, reject) => {
            const observable = this.hotResolve(route, state);
            observable.first().subscribe(() => resolve(observable));
        }
    }

    public abstract hotResolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): T;
}
```
```javascript
@Injectable()
export class DocumentsResolve extends HotResolve<Observable<IDocument[]>> {
    constructor(
        private _documentService: DocumentService
    ) {
        super();
    }

    public hotResolve(
        route: ActivatedRouteSnapshot,
        state: RouterStateSnapshot): Observable<IDocument[]> {

        return this._documentService.list();
    }
}
```

**Comment by Rodrigo54** [comment]
A good example for using this feature is the own angularfire2 - Firestore that work with real-time data updates. I would like to continue to note the updates to the resolved route data.

**Comment by tatsujb** [comment]
hey @greggbjensen, I'd like to utilize your code but without the Resolve example, I can't figure it out. could you please paste that too? thank you

**Comment by tatsujb** [comment]
hey @greggbjensen , in your example (second version) what must DocumentService be?

Is it calling the service that I'm failing to get past navigation or is it vice versa and I'm supposed to call and use DocumentResolve with my service, if so what's DocumentService, I don't get how to use your code.

**Comment by syndicatedshannon** [comment]
@greggbjensen : Actually I think it takes more boilerplate code than that. You might have it already, under the covers elsewhere, but without it, here's what probably happens:

1. observable is created
2. observable is subscribed
3. network request is made
4. first item arrives
5. subscription is released
6. observable is provided to component
7. observable is subscribed
8. network request is made
9. component is constructed with no values pending on observable stream (as though not resolved at all)

That's the reason for the reference to "PublishReplay" in my original comment. Years later, we are still fighting this issues pretty severely. We've iterated through many solutions, all of which are far from perfect due to lack of support for this need at various points in router code. For example, another problem is "where is the best place to unsubscribe?", which is why I posed [this question](https://stackoverflow.com/questions/48266343/an-event-for-cleaning-up-angular-router-resolved-data) on S/O. Yet another problem is a maintainable pattern for defining resolvers for this.

If you like, I'll try to dig up some of our current base classes, but as I mentioned, none of it is pretty (not for lack of trying).

**Comment by syndicatedshannon** [comment]
Here's the core code for our current implementation relative to this topic:

```
export abstract class ViewModelFactoryBase implements Resolve<ViewModel> {
	constructor(private router: Router) { }

	// could possibly implement a shareReplay handoff, instead of this publishReplay connection manager,
	// by monitoring ActivationEnd, which holds the ActivatedRouteSnapshot to directly unsubscribe or disconnect
	resolve(route: ActivatedRouteSnapshot, router: RouterStateSnapshot) {
		let vm = RouteValue.useActiveProvider<ViewModelFactoryBase>(route).value(this, (k) => route.paramMap.get(k));
		this.uponLeaving(route, () => vm.disconnect());
		return vm.connect();
	}

	private uponLeaving(route: ActivatedRouteSnapshot, action: () => void) {
		// TODO: This doesn't account for query parameters
		let url = [""].concat(...route.pathFromRoot
			.filter(r => r.url instanceof Array)
			.map(r => r.url.map(s => s.path)))
			.join('/');
		this.router.events
			.filter(e => (
				e instanceof NavigationEnd ||
				e instanceof NavigationCancel ||
				e instanceof NavigationError
			) && !(this.router.isActive(url, false)))
			.first()
			.subscribe(action);
	}
}

export abstract class ViewModel {
	subscription = new Subscription();
	private hasConnected = false;

	connect() {
		if (this.hasConnected) {
			return;
		}
		this.hasConnected = true;
		let observables = Object.keys(this)
			.filter(k => this[k] instanceof Observable)
			.map(k => this[k] = (this[k] as Observable<any>).publishReplay(1));
		observables.forEach(o => this.subscription.add(o.connect()));
		return Observable.combineLatest(observables).first().map(_ => this);
	}

	disconnect() {
		if (this.subscription) this.subscription.unsubscribe;
	}
}
```

**Comment by syndicatedshannon** [comment]
Note that the above implementation has some hooks to solve other router issues we've encounter as well, such as distributing magic string parameters, type-safe resolves, type-safe parent route parameters, and others. We haven't resolved pattern issues surrounding AoT and related issues. So much to say, but honestly we're all kinda burned out on this issue.

**Comment by syndicatedshannon** [comment]
@jasonaden : p.s. I just added a demonstration of where this issue is causing us a lot of grief. It also shows a couple other related issue we are trying to solve. I'd love it if you would take a look and give a little guidance, if you see better ways to accomplish any of this. We've been fighting this for a long time and could use some author guidance.

**Comment by syndicatedshannon** [comment]
@kemsky : 
"1. once router stops waiting for observable to complete we will loose possibility to do redirect before component is created." - I don't see this the same way. The current assumption is that there is only one item on observable sequence. Therefore, proceeding with navigation after the first result doesn't break most implementations. I'm not suggesting allowing configuration wouldn't capture any use-cases, it just would not be relevant to most. Also, even without global configuration, returning the observable as a promise, returning observable.last, observable.reduce, or any similar operators, all provide an easy "local configuration" approach.

"2. we need to unsubscribe at some point and it gets tricky when components are reused by the router" - totally agree. that's why this is so hard. it becomes harder because UrlTree constructors and other related tooling isn't exported for public use.

**Comment by jasonaden** [comment]
@syndicatedshannon Understood. I definitely agree this is an issue. Right now we're preparing for v6 RC.0, which means getting a few critical fixes in plus some updates to support the new rxjs.

That being said, shortly after RC.0 starts, I'll be diving in on this topic plus others with regards to the router and hopefully directly addressing how Observables are exposed/used within the router.

I know it's been a while on this topic. Part of the problem is not having consistency across the framework in terms of how Observables are exposed and used. But this is something we're looking to standardize on, and it will likely start with the Router.

**Comment by greggbjensen** [comment]
@tatsujb and @syndicatedshannon, here is an example of the DocumentService.  It holds onto the original Observable for the request that was made.  It then uses fetch to get the list again and publish back through the same stream, when an update, create, or delete is made.

Any component that calls `list()` will get the cached observable and the same stream that updates.

```typescript
// Cache for hot observable that allows stream to be cached and pushed.
export class SubjectFetch<T> {
    private _subject: ReplaySubject<T>;
    private _observable: Observable<T>;
    private _fetch: () => Observable<T>;

    constructor(fetch: () => Observable<T>) {
        this._subject = new ReplaySubject<T>();
        this._observable = this._subject.asObservable();
        this._fetch = fetch;
    }

    public fetch(aggregate: boolean = true): void {
        this._fetch()
            .subscribe(value => this._subject.next(value));
    }

    public get observable(): Observable<T> {
        return this._observable;
    }
}

@Injectable()
export class DocumentService {
    private _documentsCache: SubjectFetch<IDocument[]>;

    constructor(private _http: HttpClient) { }

    public list(): Observable<IDocument[]> {

        // Use cache if we have it.
        if (!this._documentsCache) {
            this._documentsCache = new SubjectFetch(() =>
                this._http.get<IDocument[]>(`http://somedomain.com/api/documents`));
        }

        return this._documentsCache.observable;
    }

    public create(document: IDocument): Observable<IDocument> {
        const result = this._http.post<IDocument>(
            `http://somedomain.com/api/documents`, document);

        // Fetch documents list again and publish back through hot observable.
        result.subscribe(() => this._documentsCache.fetch());

        return result;
    }
}
```

**Comment by syndicatedshannon** [comment]
@greggbjensen : do your observables all end themselves? I don't see the subscription token saved anywhere? Sorry, I'm not really following how this works. It's not critical, as I think I get the general idea relative to this ticket.

**Comment by tatsujb** [comment]
@syndicatedshannon As far as I know on services and injectables you don't need to unsub since they handle this themselves. Unsubbing is a pattern reserved for components.

**Comment by syndicatedshannon** [comment]
@jasonaden : Thanks for the reply. I'm going to try to summarize what I see as the major router/resolve-related issues/enhancements raised here. It will take me a few edits. Before I'm done I'll also try to dig up related tickets that already capture individual line items. This will help me identify which items don't actually have issues created for them yet. Also, when you do start looking at this issue, maybe this list will be helpful.

1. confusing when waiting for observable to complete appears to hang application without error
2. passing hot-resolved observable to component requires complex boiler-plate manipulation
3. unsubscribing from hot-resolved observable requires asymmetric post-resolve code
4. no event notifies when router releases resolved items (possible solution for line 3 above)
5. difficult to evaluate complete route 'identity' for comparison (alternate solution for line 3)
6. router resolve requires matching magic parameter strings between route definition and resolver
7. AoT doesn't support functions (or arrow expressions) in route definitions
8. router resolve does not readily support simultaneously parameters and dependency injection

**Comment by syndicatedshannon** [comment]
@tatsujb : that's only true when following a relatively limited set of use-cases. if the router doesn't perform the subscribe or alter the observable in some way, it won't have a token to unsubscribe from.

**Comment by syndicatedshannon** [comment]
@tatsujb : for example, I would generally trust a component to unsubscribe on an async pipe (unless it were a REALLY expensive operation, which I'd try to avoid anyway). I would also expect an observable handed directly to Resolve<T> to be unsubscribed the moment resolution is complete (precisely because resolution completion is coincident with unsubscribe). But handing resolve a `ReplaySubject` or `ConnectableObservable`? All bets are off.

**Comment by syndicatedshannon** [comment]
@tatsujb : 

That's only true when following a relatively limited set of use-cases. if the router doesn't perform the subscribe or alter the observable in some way, it won't have a token to unsubscribe from.

For example, I would generally trust a component to unsubscribe on an async pipe (unless it were a REALLY expensive operation, which I'd try to avoid anyway). I would also expect an observable handed directly to Resolve<T> to be unsubscribed the moment resolution is complete (precisely because resolution completion is coincident with unsubscribe). But handing resolve a `ReplaySubject` or `ConnectableObservable`? All bets are off.

You can test this yourself by adding `.finally(() => console.log('here'))` to your operators and see when it happens.

**Comment by greggbjensen** [comment]
@syndicatedshannon you are right, in most cases you should unsubscribe.  I updated by example above to have `first()` in both cases which automatically cleans up the subscription.

**Comment by pickfire** [comment]
It was also confused at first that `forkJoin` did not get any data from `paramMap` since it does nto complete (I used to use `take(1)` until I saw `first()` above).

```typescript
$this.observable$ = forkJoin([
  this.route.paramMap.pipe(first()),
  this.service.get()
]).pipe(d => map(...));
```

**Comment by trotyl** [comment]
Also please note the `Observable.prototype.toPromise()` do will wait for complete. So another question would be should `return data$;` and `return data$.toPromise();` in reolve have different behavior?

**Comment by salomoni** [comment]
Any updates on this?

**Comment by Airblader** [comment]
@pickfire 

> It was also confused at first that forkJoin did not get any data from paramMap since it does nto complete 

`forkJoin` doesn't even *emit* until all inner observables *completed*, so you'd be having an issue either way.

@trotyl 

> Also please note the Observable.prototype.toPromise() do will wait for complete. So another question would be should return data$; and return data$.toPromise(); in resolve have different behavior?

I don't see what these things have to do with one another. Observables are a fundamentally different concept from promises, and promises don't have any distinction of "emission" and "completion". They just resolve and that's it. It just so happens that the behavior of `toPromise` in rxjs that it waits for completion rather then emission of the observable, but that's a design matter of rxjs.

**Comment by trotyl** [comment]
> Observables are a fundamentally different concept from promises, and promises don't have any distinction of "emission" and "completion".

@Airblader That's true, and my conclusion is:

 **`Resolve` (or any other `Guard`) shouldn't support `Observable` at all**.

All of these issue are simply due to concept of `Observable` not fit into concept of `Guard`. The role of a `Guard` is only wait for a possible-async task to complete, the phasing is easy to understand in [`RouterEvent`](https://angular.io/api/router/RouterEvent):

+ [`GuardsCheckStart`](https://angular.io/api/router/GuardsCheckStart), [`GuardsCheckEnd`](https://angular.io/api/router/GuardsCheckEnd);
+ [`ResolveStart`](https://angular.io/api/router/ResolveStart), [`ResolveEnd`](https://angular.io/api/router/ResolveEnd);

There's never a temporal dynamic binding in `Guard`, making `Observable` conceptually a footgun in first place. Also all of these task indicators are function return values, not shared properties, the *definition of start* is already determined by the invocation of the function, making an extra `SUBSCRIPTION` phasing also being nonsense.

Making `Observable` acceptable simply because they've been used a lot isn't a legit approach, the same issue apply to [`loadChildren`](https://angular.io/api/router/LoadChildrenCallback), it's ridiculous to have a `() => Observable` interface which simply duplicates the definition of start, besides the temporal emission totally not making sense for code-splitting.

Having first-class support of `Observable` in Angular is great, but abusing it regardless of the semantics would be absolutely harmful.

**Comment by Airblader** [comment]
I see your argument, but I disagree because I think the 99% usecase of a guard is a simple HTTP call, which returns an observable due to the HttpClient design (where observables really are the correct choice), so forcing users to convert this to a Promise all the time is just cumbersome. So to me it's logical to return an observable from the guard. 

I also think it's perfectly reasonable if the guard waits for the first emission rather than completion, I don't have a preference for that.

**Comment by trotyl** [comment]
> but I disagree because I think the 99% usecase of a guard is a simple HTTP call, which returns an observable due to the HttpClient design

It would simply break when having the wrong `observe`, as an HTTP `Observable` doesn’t have to be single-value, making this approach error-prone.

**Comment by trotyl** [comment]
> A http call that emits multiple times is also a rare thing, and a user would probably know when they are doing that.

That’s why requiring complete being acceptable while taking the first isn’t. If you’re not making multiple-value observable than current requirement won’t affect your use case. But it’s not acceptable to make others to non-debuggable error as they’re using the API correctly.

**Comment by Airblader** [comment]
It sounds to me like you're arguing that the correct API should be waiting for the completion, or in other words that you think the current behavior shouldn't be changed at all (which is fine with me).

For completeness sake, though, if guards would use the first emission of an observable, you could still just design your observable to filter out any emission that aren't actually the result you want to give. I think that's perfectly reasonable. At the end of the day, the guard wants one value and one value only, so whether it uses the first or last emission of an observable for that is purely a design choice and doesn't prevent any particular usecase from working.

**Comment by trotyl** [comment]
> you could still just design your observable to filter out any emission that aren't actually the result you want to give. the guard wants one value and one value only

Making explicit validation of emit times (throwing error on multiple emits) is also reasonable to me, having a must-only-emit-one-value requirement is valid, as long as it’s forced by framework (rather than relying on user). That’s also why not waiting for complete isn’t valid, would even break in a `startWith` operator which is common practice in async programming.

**Comment by Airblader** [comment]
The framework cannot enforce that. If we decided to use the first emission, we'd logically unsubscribe from it once we saw that value, thus never knowing whether a second value would have been emitted or not.

It doesn't matter if guards use the first or last value: the user has to be careful either way. And either way, the framework cannot ensure correct usage or give warnings. 

If we wait for completion (current behavior), the user has to ensure that the observable completes. That's what spawned this issue in the first place.

**Comment by trotyl** [comment]
> The framework cannot enforce that.

That’s why waiting for complete must be happening.

> If we wait for completion (current behavior), the user has to ensure that the observable completes.

It can be handled by a `ROUTER_GUARD_TIMEOUT` configuration, I didn’t say current forever-waiting strategy is the right one.

**Comment by Airblader** [comment]
Perhaps a timeout that in dev-mode only prints a warning message to the console would be useful, but in production mode this timeout needs to be disabled for sure. The framework cannot dictate how long the guard is allowed to take.

**Comment by trotyl** [comment]
I agree, it might be made an option like `enableTracing` and controlled by environment.

**Comment by pickfire** [comment]
> If we wait for completion (current behavior), the user has to ensure that the observable completes. That's what spawned this issue in the first place.

I believe the user only know about that when they search for this issue online or that they found this thread, this is not easily discovered by the users and is not checked during compile time.

> Perhaps a timeout that in dev-mode only prints a warning message to the console would be useful, but in production mode this timeout needs to be disabled for sure. The framework cannot dictate how long the guard is allowed to take.

Looks interesting, it might be something like the `dev-mode` warning in the console. I bet this could avoid a lot of users from getting into this pitfall by default.

**Comment by Airblader** [comment]
> I believe the user only know about that when they search for this issue online or that they found this thread, this is not easily discovered by the users and is not checked during compile time.

Yes, because it *cannot* be checked. The framework can never know whether the user forgot to complete the observable or whether the completion will still happen.

**Comment by trotyl** [comment]
> I believe the user only know about that when they search for this issue online or that they found this thread, this is not easily discovered by the users and is not checked during compile time.

Yes, from my feeling avoid implicit failure (unexpected behavior without failing) is the first-class requirement, which should be ensured in any path.

Also, wrong timing (progressive) could be even more worse than not finishing (yes/no), so I'd suggest it always been covered by runtime check rather than relying on user reading the documentation carefully, or simply drop the support for `Observable` if the API fails to be intuitive.

**Comment by Airblader** [comment]
> simply drop the support for Observable if the API fails to be intuitive.

As I said before, killing the 99% usecase sounds like a bad idea to me. If a simple `return this.http.get()` no longer works, don't you think there'd be a constant stream of users asking "why isn't it working", "how can I make it work", "why can't I return an observable"?

If we want to deal with the problem that users don't understand how observables work, we should switch from waiting for completion to waiting for first emission. This should again be the solution that solves the majority of the problems here, as generally that is what the user wants to happen. 

Note that either way this is a breaking change and with Angular being committed to avoiding these the benefit of the breaking change should be weighed against the cost of it. Removing observable support would break almost all guards in existence — that doesn't sound like a good idea to me.

**Comment by trotyl** [comment]
> If a simple `return this.http.get()` no longer works, don't you think there'd be a constant stream of users asking "why isn't it working", "how can I make it work", "why can't I return an observable"?

First, I don't think it could happen. If this is going to happen, then it would only be deprecated first at `MAJOR VERSION + 1`, and everyone could receive a warning message. And then in `MAJOR VERSION + 3` it could be removed.

> If we want to deal with the problem that users don't understand how observables work, we should switch from waiting for completion to waiting for first emission.

With those people don't understand it, how do you know what they really wants? If only first emission should be used, then only one emission should happen (thus must wait for complete), that's a good restriction which could help people understand how guard works.

> Note that either way this is a breaking change and with Angular being committed to avoiding these the benefit of the breaking change should be weighed against the cost of it.

The request itself is also a breaking change which break application silently, but removing observable (even make it breaking way) would produce an explicit error message at very early stage (compilation) which is much better.

**Comment by Airblader** [comment]
> With those people don't understand it, how do you know what they really wants?

My personal experience, so admittedly this is biased. :-) I've seen this issue come up with coworkers and on SO and I don't think even once did the user want something other than use the first emission for the guard. 

This issue mostly arises when using either switchMap from a non-completing observable to an HTTP call (get current user, request auth status) or when using forkJoin on a non-completing observable. In the first case, the first emission is what the user wants, the second case wouldn't be covered by any of the solutions discussed here.

What would be the usecase for an observable that emits more than once but where you want to use the last emission? 

> If only first emission should be used, then only one emission should happen (thus must wait for complete)

I'm sorry, I'm not following what solution you're arguing for anymore. :-) This is the current behavior. The framework cannot know whether an observable will complete in the future — it's just conceptually not possible, because it'd require time travel. 

> that's a good restriction which could help people understand how guard works.

That's not a restriction, that is the current behavior. How does the current behavior help people understand how guards? 

> The request itself is also a breaking change which break application silently, but removing observable (even make it breaking way) would produce an explicit error message at very early stage (compilation) which is much better.

I disagree that you make an API more intuitive by making it more restrictive. A good API accepts any input which it can reasonably understand. That's the most user-friendly as it requires them to worry less about what they have to return — it just works, in the majority of cases. 

Observables and promises are both forms of input the guard can make sense of. 

By removing observable support from guards you're eliminating this particular issue by introducing the broader issue of people not knowing how they can make HTTP calls to implement a guard while forcing all other users to convert all their guards to promises for no apparent reason (they'll never see this discussion).

**Comment by swftvsn** [comment]
I still don't get why the router does not support hot routes, that is, routing is done for each emit. IF the route is already correct one, then only the route data is pushed to the already visible view.

All current guards etc. can be called - if the new emit changes something, the guard must be run again to see if the user still has permission to view etc.

And the disconnect, or unsubscribe would happen once the route ceases to be active.

**Comment by Airblader** [comment]
Personally, I very much like that idea. It would make it possible to have "real-time updates" to the authentication status as as soon as the status changes, the guard would do its job even if the route could previously be activated.

However, I think it might make sense to implement this as a different, new, guard (canRemainActive or something similar) as canActivate is more concerned with activating the route initially.

Maybe file a separate feature request?

**Comment by Airblader** [comment]
Actually scratch the part about a separate guard. I think it'd be more beneficial to have it in CanActivate directly as it avoids some weird questions and solved this issue as well as you described.

**Comment by trotyl** [comment]
> This issue mostly arises when using either switchMap from a non-completing observable to an HTTP call (get current user, request auth status).

This is already a quite dangerous scenario where there could be competing pending request (although old one will be cancelled) and it's not clear which one will be used for guard, in a specific application one may have knowledge of some requests being idempotent, but a library shouldn't make that conclusion.

> What would be the usecase for an observable that emits more than once but where you want to use the last emission?

Anything with a pending state (common when used in view), like:

```typescript
// will get { id: '1234' }
this.auth().pipe(
  startWith({ id: null })
)
```

> I'm sorry, I'm not following what solution you're arguing for anymore. :-) This is the current behavior. 

This is not about using last emit, but count the emission and report error when it happened more than once.

> That's not a restriction, that is the current behavior.

Again.

> That's the most user-friendly as it requires them to worry less about what they have to return — it just works, in the majority of cases.

A good API always works in any of the type-compatible inputs, rather than having additional semantic requirements, that's why `Observable` should not supported as it doesn't support ever-changing value. But it's hard to change due to already supported, hence the discussion.

> Observables and promises are both forms of input the guard can make sense of.

It doesn't make sense unless guard support any of the emission (not first one nor last one).

> By removing observable support from guards you're eliminating this particular issue by introducing the broader issue of people not knowing how they can make HTTP calls to implement a guard while forcing all other users to convert all their guards to promises for no apparent reason

If someone don't know how to convert `Observable` to `Promise`, I'd suggest them don't use `Observable`, it would only lead to more confusion when they do.

**Comment by Airblader** [comment]
> This is already a quite dangerous scenario where there could be competing pending request (although old one will be cancelled) and it's not clear which one will be used for guard, in a specific application one may have knowledge of some requests being idempotent, but a library shouldn't make that conclusion.

There's no pending request danger (that's what switchMap guarantees) and the framework doesn't have to make any conclusion here. Many guards will look exactly like that:

```
return this.userService.currentUser$.pipe(switchMap(user => this.authService.isAuthenticatedForResource(user)));
```

I don't see any issue with such an implementation other than the user may just have to add a `first()` due to the fact that otherwise this doesn't complete.

> This is not about using last emit, but count the emission and report error when it happened more than once.

OK, so your suggestion is to keep the behavior as-is, but just log a warning if there has been more than one emission? Yeah, that makes sense to me.

> If someone don't know how to convert Observable to Promise, I'd suggest them don't use Observable, it would only lead to more confusion when they do.

The same argument could be made for the issue itself here: if you don't understand the difference between emission and completion and that you need to complete your observable, don't use observables. It's a fair argument, but not a newbie-friendly one. :-) Angular beginners don't usually "choose" to use observables, the HttpClient more or less forces it on them.

**Comment by Airblader** [comment]
Maybe just as a recap, I think we have the following proposals on the table, right?

1. In dev mode, log a warning if the observable returned from the guard hasn't completed within a certain period of time.
2. In dev mode, log a warning if the observable returned from the guard has emitted more than once upon completion.
3. Change existing behavior to use the first emission rather than waiting for completion.
4. Remove observable support from guards
5. Add support for multiple emissions in the router by actually respecting each emission until the route is deactivated.

My 2c: I think #1 and #2 can both be helpful, and are non-breaking dev-only additions. #3–#5 are actual changes, of which I'd prefer #5, then #3, and very much dislike #4. :-)

**Comment by swftvsn** [comment]
Thanks for the recap @Airblader . One vote for option 5.

**Comment by trotyl** [comment]
> There's no pending request danger (that's what switchMap guarantees) and the framework doesn't have to make any conclusion here.

Consider the timing:

+ The guard is invoked and the `Observable` above is passed to router;
+ The first HTTP request being send;
+ `currentUser$` emits a new value;
+ The first HTTP request being cancelled and second HTTP request be made;
+ The second HTTP response received;
+ Guard check the second HTTP response result.

So that what being checked against is not the user when guard being invoked. In the application there might be no chance for `currentUser$` to change at that time, but it's about detailed business logic, from the usage it's indeed a dangerous approach.

> OK, so your suggestion is to keep the behavior as-is, but just log a warning if there has been more than one emission?

Many options are reasonable to me, except the take `the first value and then unsubscribe` one, thanks for listing above.

Imagine an API with:

```typescript
function process(arr: number[]): number {
  const value = arr[0]
  register(value)
}
```

Then I'd say it's definitely not intuitive, it would either:

+ Refine the logic and make use of each value;
+ Not accept array, instead a plain value;
+ Switching the type of `arr` from `number[]` to `[number]`;

Back to the issue here, [`Observable` and `Promise` are just temporal equivalents of `Array` and value](https://github.com/kriskowal/gtor), thus the options can be inferred. But there's no `Array`/`Tuple` typing different for `Observable`, so it cannot be made a compile-time check but runtime.

**Comment by Airblader** [comment]
> So that what being checked against is not the user when guard being invoked. In the application there might be no chance for currentUser$ to change at that time, but it's about detailed business logic, from the usage it's indeed a dangerous approach.

Point taken, I see now what you mean with how that causes the framework to make a decision it shouldn't be making.

**Comment by simeyla** [comment]
All I have to say is it took me less time to google the problem yesterday when I hadn't realized that the observable needed to complete than it did to read this thread today!

Of all the time I've spent with Angular and RxJS nuances over the past year this wasn't a big deal at all.

**Comment by doom777** [comment]
@trotyl 
> Anything with a pending state (common when used in view), like:
> 
> ```
> // will get { id: '1234' }
> this.auth().pipe(
>   startWith({ id: null })
> )
> ```

Resolvers get their data from services. Services should not modify observables to accomodate views, that's the Component's job, downstream from services and resolvers.


I vote for 3 for short term, and 5 long term.

The reason that it **is** intuitive, is that the resolver does not process the data, it just make sure there is at least some data before loading the component UI. Once there is, its job is done, and it just has to pass the data observable down to the component.

**Comment by yglin** [comment]
Stumbled upon this issue and vote for 2, then 3

But deep in heart I vote for 4,  it's better resolver stick with Promise instead of Observable.
The reason here is, IMHO, change of route state should be only triggered by change of url path, not by data stream. The responsibility of resolver, or guard, should be simply waiting, not observing.

**Comment by syndicatedshannon** [comment]
Regarding '#4' on @Airblader 's [summary](https://github.com/angular/angular/issues/10556#issuecomment-440187934) above, it is related to but does not address this issue, which is primarily about Resolve. The statement would need to be rephrased to "remove observable support from guards and resolve".

Guarding is an extra step, not about the readiness of data, even if they support Observables, but about tests on ready data. Personally, I've accepted that guards cannot 'fix-up' resolve because they are an 'if' question, not a 'when' question. I have tried to use them for such only because they seemed like an accessible development point.

I can't say with certainty, why observables were chosen for the basis for resolve and yet they do not actually support data streaming, but it's likely the designer had in mind the ability to cancel network requests, which is obviously important. I appreciate that a promise appears to provide a more intuitive return type, but only if we overlook cancellation.

At a glance, as currently phrased, #4 is a non-starter.

**Comment by syndicatedshannon** [comment]
IMO, an important aspect missing from that same [summary](https://github.com/angular/angular/issues/10556#issuecomment-440187934) is providing support for unsubscribing. For our projects, it may even be the most essential aspect, although it's hard to be certain since we've become accustomed to certain behaviors since this issue report 3 years ago.

As many have suggested (on this thread and others), we can convert the `Observable` to a `BehaviorSubject` in a resolve handler. However, there is no unique ID associated with a route resolve action to track when that subscription has been expired, so we have to write our own route management handlers as well. That itself is challenging, because much of the code that implementation depends on is not public Angular API.

Overall, I feel like, after RoutableComponents had such a rocky start, no one on the Angular team has been eager to take on these issues, which probably present not only design challenges but have historical baggage as well.

Personally, if someone wants to as simply as possibly patch this issue for me and me only, I'd be satisfied with a symmetric resolve capability: knowledge of when a route enters and exits a particular requirement.

**Comment by gabrieltbmaestro** [comment]
I've hit a bump in the road with this so far because some of my data comes from a websocket connection. What I thought I'd do is use a guard for authentication and authorization, followed by a resolver for this data, but as this thread confirms, nothing happens.

How feasible is #5? Right now, my biggest issue is on manual URL changes which triggers an asymmetrical reloading of data across my components. Does anyone have a solution for this?
Otherwise, have my vote for #5, fwiw.

**Comment by atscott** [comment]
The basic premise of resolvers is to block component instantiation until the items have been resolved. If there was no need to have the data before component instantiation, the loading logic should be moved to the component instead.

For the case where new data needs to be resolved when something changes, this is somewhat available through the [runGuardsAndResolvers](https://angular.io/api/router/Route#runGuardsAndResolvers) option, where the resolver data could be updated when parameters change.

Other use-cases may generally be covered by #42953 if the router were more extensible and allowed for more configuration as to how guards and resolvers are run. 

Closing in favor of #42953

**Comment by syndicatedshannon** [comment]
Complete not equal resolve

**Comment by gjensen-gomeyra** [comment]
@atscott I do agree the premise of resolvers is to have the data before component instantiation.  Which is still true in my example.  However, there are many cases where I want to receive additional updates to the data my components can respond to.  This is what I use the HotResolver for.  It verifies initial data was loaded and I am able to continue to listen to that data for changes.  This still seems like a very valid use of a resolver to me.  Initial data should be required, but that does not mean the observable has to be completed.

Maybe the change here would be to simply use first() on a resolver instead of requiring a completed observable.

**Comment by greggbjensen** [comment]
@atscott I do have it working using a wrapper, as shown above.   I just thought it would be useful to other users.  I can also create a library that developers can add to their projects.  Maybe that would be the right route to go.

**Comment by syndicatedshannon** [comment]
@greggbjensen 's case is my main case as well. Similar to @greggbjensen , I've also already integrated a workaround for this.

> The basic premise of resolvers is to block component instantiation until the items have been resolved. If there was no need to have the data before component instantiation, the loading logic should be moved to the component instead.

@atscott 's closing comment didn't demonstrate understanding of this specific issue/thread, but the follow-up to @greggbjensen on July 26 did. Thank both of you for that.

I do appreciate that Observable was chosen for HTTP calls because of the ability to chain cancellations and similar, and router today basically expects it to behave as a cancellable promise. Therefore, I do see how this can be considered "as-designed".

It's very confusing in the current architecture to route hot subscription data into components, and it seems like a common feature to provide. That is the main thrust of this issue. I hope #42953 allows for that, whatever its design entails.
